### PR TITLE
(fix): sparse indexing

### LINKED ISSUE #11290: `2024.8.0` breaks `sparse` array indexing
**Describe the issue**:

Previously, one could at least use `dask` for sparse array indexing/getting, even if more complex operations didn't work.  But now, this too has broken, at least in one use-case.

**Minimal Complete Verifiable Example**:
The following works on `<2024.8.0` but breaks otherwise
```python
import dask.array as da
import scipy as sp

X = da.map_blocks(lambda: sp.sparse.random(10, 10, format="csr"), meta=sp.sparse.random(10, 10, format="csr"), chunks=((10, ) * 100, (10, )))

X[[5, 15], :].compute()
```

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[11], line 1
----> 1 X[[5, 15],:].compute()

File ~/Projects/Theis/anndata/venv/lib/python3.11/site-packages/dask/base.py:376, in DaskMethodsMixin.compute(self, **kwargs)
    352 def compute(self, **kwargs):
    353     """Compute this dask collection
    354 
    355     This turns a lazy Dask collection into its in-memory equivalent.
   (...)
    374     dask.compute
    375     """
--> 376     (result,) = compute(self, traverse=False, **kwargs)
    377     return result

File ~/Projects/Theis/anndata/venv/lib/python3.11/site-packages/dask/base.py:662, in compute(traverse, optimize_graph, scheduler, get, *args, **kwargs)
    659     postcomputes.append(x.__dask_postcompute__())
    661 with shorten_traceback():
--> 662     results = schedule(dsk, keys, **kwargs)
    664 return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])

File ~/Projects/Theis/anndata/venv/lib/python3.11/site-packages/dask/array/_shuffle.py:190, in concatenate_arrays(arrs, sorter, axis)
    189 def concatenate_arrays(arrs, sorter, axis):
--> 190     return np.take(np.concatenate(arrs, axis=axis), np.argsort(sorter), axis=axis)

ValueError: zero-dimensional arrays cannot be concatenated
```
**Anything else we need to know?**:
I've noticed that the problem only occurs when indexing across the chunk boundaries, for what it is worth.
**Environment**:

- Dask version: 2024.8.0
- Python version: 3.11
- Operating System: mac
- Install method (conda, pip, source): pip

### PR DESCRIPTION
**Pull request by ilan-gold**
- [x] Closes #11290
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

cc: @phofl

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

Admins can comment `ok to test` to allow this one PR to run or `add to allowlist` to allow all future PRs from the same author to run.

File: dask/array/_shuffle.py [hunk 1]
```diff
@@ -294,8 +294,30 @@ def _getitem(obj, index):
 
 
 def concatenate_arrays(arrs, sorter, axis):
-    concatenate = concatenate_lookup.dispatch(type(arrs[0]))
-    return np.take(concatenate(arrs, axis=axis), np.argsort(sorter[1]), axis=axis)
+    try:
```
Comment (phofl) [inline dask/array/_shuffle.py:297]: You have to hook this into our dispatching mechanism, cc @rjzamora who will probably be able to provide guidance here

Comment (ilan-gold) [reply dask/array/_shuffle.py:297]: Thanks @phofl - yeah, I'm not exactly sure what dispatching mechanism you're referring to.  If you want me to create one for indexing, I can (like `concatenate_lookup` and `tensordot_lookup`)

Comment (phofl) [reply dask/array/_shuffle.py:297]: Yeah we basically need something like ``concatenate_lookup`` for this

Comment (ilan-gold) [reply dask/array/_shuffle.py:297]: Great will do

Comment (phofl) [reply dask/array/_shuffle.py:297]: maybe a better name than ``concatenate_arrays`` if you end up defining the function outside of ``_shuffle.py``

Comment (ilan-gold) [reply dask/array/_shuffle.py]: I guess what I'm saying is that something like a `concatenate_and_take` registray would need to some include a way of dispatching the `take` which is what I do here.  But we wouldn't need a registry for that then.  I'll think on it some more.

File: dask/array/_shuffle.py [hunk 2]
```diff
@@ -294,8 +294,10 @@ def _getitem(obj, index):
 
 
 def concatenate_arrays(arrs, sorter, axis):
-    concatenate = concatenate_lookup.dispatch(type(arrs[0]))
-    return np.take(concatenate(arrs, axis=axis), np.argsort(sorter[1]), axis=axis)
+    typ = type(arrs[0])
+    take = take_lookup.dispatch(typ)
+    concatenate = concatenate_lookup.dispatch(typ)
+    return take(concatenate(arrs, axis=axis), np.argsort(sorter[1]), axis=axis)
```
Comment (rjzamora) [inline dask/array/_shuffle.py:300]:
```suggestion
    return take_lookup(
        concatenate_lookup(arrs, axis=axis),
        np.argsort(sorter[1]),
        axis=axis,
    )
```

File: dask/array/_shuffle.py (hunk 1 above)
Comment (rjzamora) [reply dask/array/_shuffle.py]: As far as I can tell, the `take_lookup` solution makes sense. It's probably best to dispatch on primitive logic that is likely to be used in many places (e.g. concat and take).

File: dask/array/_shuffle.py (hunk 2 above)
Comment (ilan-gold) [reply dask/array/_shuffle.py:300]: @rjzamora these are lists so the `__call__` will not work for `concatenate_lookup`

Comment (ilan-gold) [reply dask/array/_shuffle.py]: but will do for `take_lookup`

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
thanks

File: dask/array/_shuffle.py (hunk 2 above)
Comment (rjzamora) [reply dask/array/_shuffle.py]: Ah, sorry - My mistake. I was probably thinking of the [dataframe `concat` function](https://github.com/dask/dask/blob/091f4e9bd226836faeaf30bdc3b5182ba14d22a6/dask/dataframe/dispatch.py#L34) - We could add a similar function to dask.array, but I think the way you did things is fine.

**Comment by ilan-gold** [comment]
my pleasure!

**Pull request merged**